### DISCUSSION TITLE
Circular Progressindicator

### DISCUSSION THREAD

**Discussion by frankalex92** [slint-ui/slint#4656] [Q&A]
Hi,

I was wondering if it´s possible to create a Progressindicator that moves in a circular motion like this:

![image](https://github.com/slint-ui/slint/assets/160749116/e194ec8e-dd88-4cc5-a5f0-c41a6d874492)

Thanks for your help!

BR
Alex

**Comment by ogoffart** [comment]
You can use the `Path` element to draw an ark like that.

For example:
```
export component CircularProgress {
    preferred-height: 100%;
    preferred-width: 100%;
    in property <float> progress;

    path := Path {
        private property <float> radius: 0.5;
        // clamp is a workaground to get filled circle by 1.0
        private property <angle> progress: clamp(root.progress * 1turn, 0, 0.99999turn);

        viewbox-width: 1;
        viewbox-height: 1;
        width: 100%;
        height: 100%;

        stroke-width: 10px;
        stroke: Palette.accent-background;

        MoveTo {
            x: 0.5;
            y: 0;
        }

        ArcTo {
            radius-x: path.radius;
            radius-y: path.radius;
            x: 0.5 - path.radius * sin(-(path.progress) );
            y: 0.5 - path.radius * cos(-(path.progress) );
            sweep: root.progress > 0;
            large-arc: root.progress > 0.5;
        }
       
    }

    text := Text {
        font-size: parent.width / 4;
        color: Palette.accent-background;
        text: round(progress * 100);
    }
    Text {
        font-size: parent.width / 8;
        y: text.y + text.height - self.height - self.font-size/4;
        x: text.x + text.width;
        text: "%";
    }
}
```

[SlintPad link](https://slintpad.com/?snippet=import+%7B+AboutSlint%2C+Button%2C+VerticalBox%2C+Palette%2C+Slider+%7D+from+%22std-widgets.slint%22%3B%0A%0A%0Aexport+component+CircularProgress+%7B%0A++++preferred-height%3A+100%25%3B%0A++++preferred-width%3A+100%25%3B%0A++++in+property+%3Cfloat%3E+progress%3B%0A%0A++++path+%3A%3D+Path+%7B%0A++++++++private+property+%3Cfloat%3E+radius%3A+0.5%3B%0A++++++++%2F%2F+clamp+is+a+workaground+to+get+filled+circle+by+1.0%0A++++++++private+property+%3Cangle%3E+progress%3A+clamp%28root.progress+*+1turn%2C+0%2C+0.99999turn%29%3B%0A%0A++++++++viewbox-width%3A+1%3B%0A++++++++viewbox-height%3A+1%3B%0A++++++++width%3A+100%25%3B%0A++++++++height%3A+100%25%3B%0A%0A++++++++stroke-width%3A+10px%3B%0A++++++++stroke%3A+Palette.accent-background%3B%0A%0A++++++++MoveTo+%7B%0A++++++++++++x%3A+0.5%3B%0A++++++++++++y%3A+0%3B%0A++++++++%7D%0A%0A++++++++ArcTo+%7B%0A++++++++++++radius-x%3A+path.radius%3B%0A++++++++++++radius-y%3A+path.radius%3B%0A++++++++++++x%3A+0.5+-+path.radius+*+sin%28-%28path.progress%29+%29%3B%0A++++++++++++y%3A+0.5+-+path.radius+*+cos%28-%28path.progress%29+%29%3B%0A++++++++++++sweep%3A+root.progress+%3E+0%3B%0A++++++++++++large-arc%3A+root.progress+%3E+0.5%3B%0A++++++++%7D%0A+++++++%0A++++%7D%0A%0A++++text+%3A%3D+Text+%7B%0A++++++++font-size%3A+parent.width+%2F+4%3B%0A++++++++color%3A+Palette.accent-background%3B%0A++++++++text%3A+round%28progress+*+100%29%3B%0A++++%7D%0A++++Text+%7B%0A++++++++font-size%3A+parent.width+%2F+8%3B%0A++++++++y%3A+text.y+%2B+text.height+-+self.height+-+self.font-size%2F4%3B%0A++++++++x%3A+text.x+%2B+text.width%3B%0A++++++++text%3A+%22%25%22%3B%0A++++%7D%0A%7D%0A%0A%0A%0Aexport+component+Demo+%7B%0A++++VerticalBox+%7B%0A++++++++alignment%3A+start%3B%0A++++++++Text+%7B%0A++++++++++++text%3A+%22Hello+World%21%22%3B%0A++++++++++++font-size%3A+24px%3B%0A++++++++++++horizontal-alignment%3A+center%3B%0A++++++++%7D%0A+++++++%0A++++++++HorizontalLayout+%7B+alignment%3A+center%3B+Button+%7B+text%3A+%22OK%21%22%3B+%7D+%7D%0A++++++++progress+%3A%3D+CircularProgress+%7B%0A++++++++++++height%3A+self.width%3B%0A++++++++++++progress%3A+slider.value+*+1%25%3B%0A++++++++%7D%0A++++++++slider+%3A%3D+Slider+%7B%0A++++++++++++value%3A+75%3B%0A++++++++++++maximum%3A+100%3B%0A++++++++%7D%0A++++%7D%0A%7D%0A&style=native)

**Comment by frankalex92** [comment]
Ah, don't know why I assumed that something drawn with the path feature can´t change during runtime..should´ve thought of that.

Thanks!
